File: samples-scala/src/test/scala/io/github/ilaborie/loops/samples/AccumulateTest.scala
package io.github.ilaborie.loops.samples

import org.junit.Assert.assertEquals
import org.junit.Test

class AccumulateTest {


  private def test(func: List[ElementScala] => Int): Unit = {
    val elements = List(
      ElementScala("1"),
      ElementScala("_2"),
      ElementScala("__3")
    )
    val result = func(elements)
    assertEquals(6, result.toLong)
  }

  @Test
  def for_should_accumulate(): Unit = {
    test {
      io.github.ilaborie.loops.samples.anatomy.AccumulateScala.accumulate
    }
  }

  @Test
  def recursion_should_accumulate(): Unit = {
    test {
      io.github.ilaborie.loops.samples.recursion.AccumulateScala.accumulate
    }
  }

  @Test
  def recursion_should_accumulate_alt(): Unit = {
    test {
      io.github.ilaborie.loops.samples.recursion.AccumulateScala.accumulateAlt
    }
  }

  @Test
  def tail_recursion_should_accumulate(): Unit = {
    test {
      io.github.ilaborie.loops.samples.recursion.AccumulateTrScala.accumulate
    }
  }

  @Test
  def tail_recursion_should_accumulate_alt(): Unit = {
    test {
      io.github.ilaborie.loops.samples.recursion.AccumulateTrScala.accumulateAlt
    }
  }

  @Test
  def stream_should_accumulate(): Unit = {
    test {
      io.github.ilaborie.loops.samples.stream.AccumulateScala.accumulate
    }
  }

}

File: samples-kotlin/src/test/kotlin/io/github/ilaborie/loops/samples/stream/AccumulateTest.kt
package io.github.ilaborie.loops.samples.stream

import io.github.ilaborie.loops.samples.ElementKt
import org.junit.Assert.assertEquals
import org.junit.Test

class AccumulateTest {

    private fun test(func: (List<ElementKt>) -> Int) {
        val elements = listOf(
            ElementKt("1"),
            ElementKt("_2"),
            ElementKt("__3")
        )
        val result = func(elements)
        assertEquals(6, result.toLong())
    }

    @Test
    fun for_should_accumulate() {
        test { io.github.ilaborie.loops.samples.anatomy.accumulate(it) }
    }

    @Test
    fun recursion_should_accumulate() {
        test { io.github.ilaborie.loops.samples.recursion.accumulate(it) }
    }

    @Test
    fun recursion_should_accumulate_alt() {
        test { io.github.ilaborie.loops.samples.recursion.accumulateAlt(it) }
    }

    @Test
    fun tail_recursion_should_accumulate() {
        test { io.github.ilaborie.loops.samples.recursion.accumulateTR(it) }
    }
    @Test
    fun tail_recursion_should_accumulate_alt() {
        test { io.github.ilaborie.loops.samples.recursion.accumulateTRAlt(it) }
    }

    @Test
    fun stream_should_accumulate() {
        test { io.github.ilaborie.loops.samples.stream.accumulate(it) }
    }

    @Test
    fun stream_should_accumulate_alt() {
        test { io.github.ilaborie.loops.samples.stream.accumulateAlt(it) }
    }

}
